### PR TITLE
[GLM5.2] elementwise fusion around sprase attention

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Fuse the pointwise work around DCP sparse attention

Three changes on the TP+DCP sparse-MLA path, each replacing a group of pointwise launches with one kernel or removing them outright. Measured on GLM-5.2 MXFP4, TP4+DCP4, MI355X. The path is shared by every `deepseek_v3`-family model (`deepseek_v3`, `kimi_k2`, `glm_moe_dsa`, `deepseek_mtp`), so this is not GLM-specific, but GLM-5.2 is what the numbers come from.

| | prefill (4451 ms) | decode (8934 ms) |
|---|---:|---:|
| affected kernels, before | 471,051 us (10.58%) | 443,372 us (4.96%) |
| after | 288,009 us (6.77%) | 254,065 us (2.97%) |
| | **-183,042 us** | **-189,307 us** |

Control kernels move ~1% across the same pair of runs (MoE gemm1 +0.4%, MLA attn +0.6%), so the deltas are clear of the noise. Numbers come from `atom.examples.profile_offline` (ISL=49152 bs=1 for prefill, ISL=4096 OSL=256 bs=48 for decode), aggregated over every launch in the trace rather than per layer — the a2a change is ~12 us in a single layer and gets swamped by run-to-run drift at that sample size.

## 1. LSE mask: six kernels → one

**Where:** `_forward_prefill_mla`, after the MLA kernel and LSE reduce, before the cross-rank combine. Under DCP the sparse path zeroes the output rows whose per-head LSE is not finite, so a row that saw no valid KV does not pollute the merge.

**Before** — `o = torch.where(torch.isfinite(lse).unsqueeze(-1), o, torch.zeros_like(o))`, which torch expands into six launches:

| kernel | what it is | us/layer |
|---|---|---:|
| `vectorized_elementwise_kernel<AbsFunctor>` | `isfinite` step 1 | 5.6 |
| `vectorized_elementwise_kernel<BinaryFunctor>` (ne) | step 2 | 5.6 |
| `vectorized_elementwise_kernel<BinaryFunctor>` (eq) | step 3 | 5.6 |
| `vectorized_elementwise_kernel<AUnaryFunctor>` (mul) | step 4 | 5.5 |
| `vectorized_elementwise_kernel<FillFunctor<float>>` | `zeros_like` | 140.2 |
| `elementwise_kernel_manual_unroll<gpu_kernel_impl_nocast>` | the select | 553.4 |

**After** — `_zero_nonfinite_rows_kernel` (Triton, new in `dcp_ops.py`), **281.5 us/layer, -61%**.

The four mask kernels are pure launch overhead: the mask is `[B, H]`, a few KB, and each launch costs ~5.5 us. The select is the expensive one and the reason a kernel beats a better-spelled expression — it reads and rewrites every byte of `o`, while a non-finite LSE is an edge case and nearly every row survives untouched. The Triton kernel returns before touching `o` for a finite row.

Two earlier attempts are in the commit message rather than the tree: `masked_fill_` with a broadcast mask removed the fill but its kernel is slower than the select it replaced (+10.1 ms over the run), and `torch.where` with a scalar `other` got to 434 us but is still five launches.

## 2. a2a combine: two kernels → one

**Where:** the DCP all-to-all combine that merges each rank's partial attention output, between the attention output and `o_proj`.

**Before:** `_dcp_a2a_unpack_combine_kernel` wrote bf16, then `dynamic_per_token_scaled_quant_kernel<..., Li16>` immediately requantised it for `o_proj`.

**After:** `_dcp_a2a_unpack_combine_quant_kernel` emits the quantised activation directly. Whole-trace, decode:

| | launches | us |
|---|---:|---:|
| `_dcp_a2a_unpack_combine_kernel` | 20,124 → 0 | 149,236 → 0 |
| `dynamic_per_token_scaled_quant` (Li16) | 20,898 → 774 | 117,076 → 3,275 |
| `_dcp_a2a_unpack_combine_quant_kernel` | 0 → 20,124 | 0 → 123,240 |
| | | **266,312 → 126,515 (-52%)** |

Costs occupancy — the grid drops from `(tokens, heads)` to `(tokens,)` because a per-token scale spans every head — and shifts numerics by up to one bf16 step, since the fused path quantises from the fp32 accumulator instead of a bf16 round trip. Both are why it stays behind `ATOM_DCP_A2A_FUSED_QUANT`, default off. A test pins the numerics change against an fp32 torch reference.

## 3. Sparse indexer: drop the -inf prefill

**Where:** `sparse_attn_indexer` in `deepseek_v2.py`, in the chunk loop that scores the gathered KV before the per-row top-k.

**Before:** aiter's `fp8_mqa_logits` allocates `[rows, kv_aligned]` fp32 and, under its default `clean_logits=True`, writes -inf into every element so a position outside a row's `[cu_start, cu_end)` window cannot win the top-k — `vectorized_elementwise_kernel<FillFunctor<float>>`, 40 launches and 9.1 ms of a 1.3 s prefill.

**After:** no kernel. This one is a deletion, not a fusion — the work stops happening rather than moving somewhere else.

Nothing reads those positions. `top_k_per_row_prefill` is handed the same `row_starts`/`row_ends`, and every kernel in `topk_per_row_kernels.cu` offsets its reads by `rowStart` and bounds them by `rowEnd - rowStart`. `vectorized_process` aligns forward and guards head and tail scalars, so a wide load never crosses the end either. The buffer has exactly one consumer in this function, and the other five `fp8_mqa_logits` call sites in the tree are untouched.

Verified rather than argued, since a silent wrong-KV selection is the failure mode: filling everything outside the window with `+1e30` (a value that wins any top-k it is read into) and checking that no returned index lands outside the window. 14 shapes, `kv_len` deliberately off a 256 multiple (4097, 8191, 16385, 3000) to leave a real alignment tail, across the one-block path, the multi-block path (forced via `TOPK_FORCE_PATH=m`, since MI355X dispatch never picks it above batch 128) and `stable=True`. Zero out-of-window indices in every case; selected sets identical to the -inf reference wherever the kernel is deterministic at all. The one route that would read the alignment tail — calling without `rowStarts`/`rowEnds`, where `rowEnd` falls back to `len` = `stride0` = the padded width — is not reachable: both are required arguments in the Python signature.

## Correctness

A 45,372-token greedy generation (temperature 0, `ignore_eos`, TP4/DCP4) produces bit-identical tokens in all four combinations of the two env switches.

GSM8K 20-shot, `num_concurrent=64`, chat template, temperature 0:

| | flexible-extract | strict-match |
|---|---:|---:|
| baseline | 0.9682 ± 0.0048 | 0.9689 ± 0.0048 |
| §3 only (`CLEAN_LOGITS=0`) | 0.9674 ± 0.0049 | 0.9682 ± 0.0048 |
| §2 only (`FUSED_QUANT=1`) | 0.9644 ± 0.0051 | 0.9651 ± 0.0051 |
| both | 0.9621 ± 0.0053 | 0.9629 ± 0.0052 |

The baseline matches the published DCP4 reference (0.9689 / 0.9697), so the harness is sound. Isolating the switches puts the movement where §2's numerics shift predicts it: dropping the -inf prefill costs 0.07pp (one question out of 1319, well inside noise), the fused quant costs 0.38pp, and together 0.61pp. Each arm is a single run and the largest gap is ~1.2σ, so none of it is individually significant — but the ordering is consistent, and it is why `ATOM_DCP_A2A_FUSED_QUANT` stays off by default while §1 and §3 do not need a guard.

29 new tests. `pytest -k "mla or dcp or attention or prefill"` goes from 785 passed on `c61fbfb0` to 814; the 13 failures and 1 collection error are present on `c61fbfb0` too (sglang/rtpllm plugin, unrelated).

### CODE REVIEW

File: atom/model_ops/dcp_ops.py [hunk 1]
```diff
@@ -443,12 +508,111 @@ def _dcp_a2a_unpack_combine_kernel(
...
+    lse_max = tl.max(lse, axis=0)
+    lse_max = tl.where(lse_max == float("-inf"), 0.0, lse_max)
+    global_lse = tl.log(tl.sum(tl.exp(lse - lse_max), axis=0)) + lse_max
+
+    factor = tl.exp(lse - global_lse[None, :])
+    factor_is_nan = factor != factor  # noqa: PLR0124
+    factor = tl.where(factor_is_nan | (~valid[:, None]), 0.0, factor)
+
+    # [N, H, D]
+    vals = tl.load(hbase[:, :, None] + d[None, None, :]).to(tl.float32)
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py]: mask added

File: atom/model_ops/attention_mla.py [hunk 2]
```diff
@@ -1072,13 +1104,27 @@ def _dcp_project_merge_out(
             o = self._v_up_proj(
                 o, self.W_V_dcp, self.W_V_dcp_scale, num_heads=o.shape[1]
             )
+        # Only PBM hands the merge output straight to o_proj, so it is the only
+        # path whose activation quant the combine can absorb. The fp32 merge is
+        # a prefill-accuracy path and keeps its bf16 store.
+        fused_quant = None if merge_in_fp32 else self._dcp_fused_quant_dtype()
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/attention_mla.py:1142]: it's a misunderstanding due to pr description not updated. The optimization is default and no envs control is added now.

File: atom/model_ops/dcp_ops.py [hunk 3]
```diff
@@ -302,6 +302,71 @@ def _lse_pack_slots(dtype: torch.dtype) -> int:
...
+    cannot know that: it reads and rewrites all of `o` regardless. Here a
+    finite row exits before touching `o` at all, so the common case costs one
+    scalar load per row instead of a full read-modify-write of [B, H, D].
+    """
+    b = tl.program_id(axis=0).to(tl.int64)
+    h = tl.program_id(axis=1).to(tl.int64)
+
+    lse = tl.load(lse_ptr + b * lse_stride_b + h * lse_stride_h)
+    # isfinite: NaN fails `lse == lse`, +-inf fails the abs bound.
+    if (lse == lse) and (tl.abs(lse) < float("inf")):  # noqa: PLR0124
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py]: use "&" instread

File: tests/test_prefill_mla_lse_mask.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+# The kernel is Triton and the tests drive it on real tensors; both are absent
+# on the non-GPU CI runner.
+pytest.importorskip("triton")
+
+if not torch.cuda.is_available():
+    pytest.skip(
+        "compares a Triton kernel against its reference; needs a real GPU",
+        allow_module_level=True,
+    )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply tests/test_prefill_mla_lse_mask.py]: fixed

File: tests/test_dcp_a2a_fused_quant.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+import pytest
+import torch
+
+# Order matters: importing triton without an active GPU driver raises rather
+# than failing the import, so the device check has to come first.
+if not torch.cuda.is_available():
+    pytest.skip(
+        "drives Triton kernels on real tensors; needs a real GPU",
+        allow_module_level=True,
+    )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply tests/test_dcp_a2a_fused_quant.py:19]: Removed them instead. They mirrored the predicate's five ifs against a stub that hard-coded the very o_proj assumption they would need to check, so they could only confirm the code agreed with itself — and they never ran in CI anyway. The kernel tests cover this path on real tensors.

[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply tests/test_dcp_a2a_fused_quant.py:19]: Removed them instead. They mirrored the predicate's five ifs against a stub that hard-coded the very o_proj assumption they would need to check, so they could only confirm the code agreed with itself — and they never ran in CI anyway. The kernel tests cover this path on real tensors.

**Comment by ZhiweiYan-96** [comment]
> ### 🔵 Needs a closer look
> Address the FP8 scaling compatibility issue and add production-path integration coverage.
> 
> Review details
> ### Suppressed comments (2)
> **atom/model_ops/attention_mla.py:1125**
> 
> * The added quant tests invoke `_dcp_a2a_unpack_combine_quant_kernel` directly, but no test exercises this production gate and tuple handoff. A regression in `_dcp_fused_quant_dtype` could leave the standalone quant launch in place while all kernel tests still pass, and a wrong `o_proj` scale shape would only appear in the model path. Add a focused test around `_dcp_project_merge_out` covering the enabled PBM+A2A+per-token-FP8 case and the fallback combinations.
> 
> ```
>         if fused_quant is not None:
>             o, o_scale = o
>             if o_scale is not None:
>                 return self.o_proj(
>                     o.reshape(-1, self.num_heads * self.v_head_dim), o_scale
>                 )
> ```
> 
> **atom/model_ops/dcp_ops.py:609**
> 
> * The fused path computes the FP8 scale with Triton's direct `fdiv`, but the existing AITER per-token quantizer uses a precomputed reciprocal because ROCm Triton division can be one ULP different and move values to an adjacent FP8 code (`atom/model_ops/kimi_k3/quant.py:80-91`). This adds an avoidable numerical change on top of the intentional BF16-round-trip change, and the current tolerance-based test will not detect it. Pass a precomputed reciprocal into this kernel and use the established multiply/inverse-scale sequence (including the quantization operation) so this replacement remains compatible with the `o_proj` quantizer.
> 
> ```
>     scale = row_max / FP8_MAX
>     # An all-empty row gives row_max == 0; a zero scale would make the divide
>     # produce NaN and would also be an invalid dequant multiplier downstream.
>     scale = tl.where(scale > 0.0, scale, 1.0)
> ```
> 
> * **Files reviewed:** 5/5 changed files
> * **Comments generated:** 0 new
> * **Review effort level:** Lite

dcp_ops.py:609 — Fixed. Measured first: the divide gave 54% of rows a different scale and moved 0.116% of elements to an adjacent FP8 code versus get_hip_quant(per_Token). Now multiply-by-reciprocal throughout, and non-zero rows are bit-identical to aiter. The new test compares against aiter directly and fails on the old form.

attention_mla.py:1125 — This needs a real o_proj, PBM, the a2a backend and a DCP group larger than one, so it's an integration test; a stub one would only restate the code. Covered end to end instead: the offline profile shows 2652 launches of the fused kernel and zero of the unfused one, and GSM8K plus a 45k-token generation run the whole model through it — a missed tuple unpack or wrong scale shape wouldn't survive that.

File: atom/model_ops/dcp_ops.py [hunk 6]
```diff
@@ -462,10 +651,23 @@ def cp_lse_a2a(
...
     """
     n_ranks = cp_group.world_size
     if n_ranks == 1:
+        if quant_dtype is not None:
+            # Nothing to combine, so there is no kernel to fold the quant into.
+            return cp_attn_out, None
         return (cp_attn_out, cp_attn_lse) if return_lse else cp_attn_out
+    assert (
+        quant_dtype is None or not return_lse
+    ), "cp_lse_a2a: the fused-quant combine does not emit LSE"
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py]: Real bug, fixed.

File: atom/model_ops/dcp_ops.py [hunk 7]
```diff
@@ -443,12 +520,124 @@ def _dcp_a2a_unpack_combine_kernel(
...
+    d = tl.arange(0, HEAD_DIM)
+    valid = n < N_RANKS
+
+    # [N, H] base of each (shard, head) record for this token.
+    hbase = (
+        recv_ptr
+        + n[:, None].to(tl.int64) * recv_stride_n
+        + b * recv_stride_b
+        + h[None, :].to(tl.int64) * recv_stride_h
+    )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py:580]: Added the clamp; it's free and the bit-for-bit comparison against aiter still passes, so nothing changed numerically. For the record I don't think the premise holds: masking the load is the documented Triton contract, and the canonical tutorial forms out-of-range pointers for masked lanes exactly this way — as does the upstream _dcp_a2a_unpack_combine_kernel in this same file for its LSE load, which has been shipping. I also couldn't find the index clamping you attribute to the kimi_k3 kernels. Taking it as belt and braces rather than a fix.

File: atom/model_ops/dcp_ops.py [hunk 8]
```diff
@@ -302,6 +307,74 @@ def _lse_pack_slots(dtype: torch.dtype) -> int:
...
+        return
+
+    base = out_ptr + b * out_stride_b + h * out_stride_h
+    for off in tl.range(0, HEAD_DIM, BLOCK):
+        idx = off + tl.arange(0, BLOCK)
+        tl.store(
+            base + idx,
+            tl.zeros([BLOCK], dtype=out_ptr.dtype.element_ty),
+            mask=idx < HEAD_DIM,
+        )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py:355]: fixed

File: atom/model_ops/dcp_ops.py [hunk 9]
```diff
@@ -443,12 +525,129 @@ def _dcp_a2a_unpack_combine_kernel(
...
+    #     (measured: 4869 of 4.2M, and 54% of rows get a different scale).
+    #     `atom/model_ops/kimi_k3/quant.py` carries the same note.
+    #   * one reciprocal, then a multiply per element -- not a divide per element.
+    #   * the row max floored at 1e-10, which is what gives an all-zero row
+    #     aiter's tiny scale rather than a zero or a one.
+    row_max = tl.maximum(tl.max(tl.abs(acc)), AMAX_EPS)
+    scale = row_max * (1.0 / FP8_MAX)
+    tl.store(out_scale_ptr + b, scale)
+
+    q = acc * (1.0 / scale)
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/dcp_ops.py]: fixed

File: atom/models/deepseek_v2.py [hunk 10]
```diff
@@ -1869,6 +1869,14 @@ def sparse_attn_indexer(
...
                     cu_starts=row_starts,
                     cu_ends=row_ends,
+                    # The -inf prefill of this buffer has no reader. It exists
+                    # so a position outside a row's window cannot win the
+                    # top-k, but `top_k_per_row_prefill` below is handed the
+                    # same row_starts / row_ends and offsets every access by
+                    # rowStart, bounded by rowEnd - rowStart, so it never looks
+                    # outside the window. 449 us per full-index layer at
+                    # ISL=49152.
+                    clean_logits=False,
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/models/deepseek_v2.py:1879]: Checked it in: tests/test_indexer_logits_alignment_tail.py

**Comment by ZhiweiYan-96** [comment]
image

**Review by valarLip — APPROVED** [review]
